File: search-a-2d-matrix/search-a-2d-matrix.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        int m = matrix.length - 1, n = matrix[0].length - 1;
        int col = binarySearchRow(matrix, target, 0, m);
        return binarySearchCol(matrix, target, col, 0, n);
    }

    private boolean binarySearchCol(int[][] matrix, int target, int row, int low, int high) {
        while (low <= high) {
            int mid = low + (high - low) / 2;
            if (matrix[row][mid] == target) {
                return true;
            } else if (matrix[row][mid] < target) {
                low = mid + 1;
            } else {
                high = mid - 1;
            }
        }
        return false;
    }

    private int binarySearchRow(int[][] matrix, int target, int low, int high) {
        int n = matrix[0].length - 1;
        while (low < high) {
            int mid = low + (high - low) / 2;
            if (matrix[mid][n] == target) {
                return mid;
            } else if (matrix[mid][n] < target) {
                low = mid + 1;
            } else {
                high = mid;
            }
        }
        return low;
    }
}

File: search-a-2d-matrix/search-a-2d-matrix.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        row = self.findRow(matrix, target)
        return self.findCol(matrix, row, target)


    def findRow(self, matrix: List[List[int]], target: int) -> int:
        low, high = 0, len(matrix) - 1
        while low < high:
            mid = int(low + (high - low) / 2)
            if (matrix[mid][len(matrix[0])-1] < target):
                low = mid + 1
            else:
                high = mid
        return low

    def findCol(self, matrix: List[List[int]], row: int, target: int) -> bool:
        low, high = 0, len(matrix[0]) - 1
        while low <= high:
            mid = int(low + (high - low) / 2)
            if matrix[row][mid] == target:
                return True
            elif matrix[row][mid] < target:
                low = mid + 1
            else:
                high = mid - 1
        return False